community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -626,7 +623,17 @@ private void afterRollback()
     private void release()
     {
         locks.releaseAll();
-        pool.release( this );
+        if ( terminated )
+        {
+            // This transaction has been externally marked for termination.
+            // Just dispose of this transaction and don't return it to the pool.
+            dispose();
+        }
+        else
+        {
+            // Return this instance to the pool so that another transaction may use it.
+            pool.release( this );
```

Comment:
For the time being that might be OK, but mismatches a bit with the Locks.Client handed to the KTI instance in the constructor. Disposing the whole transaction is just safer all around. Master switches happen rarely compared to most other things and so the overhead of disposing the transaction instances is negligible.